ui/app/hooks/useEqualityCheck.js

```diff
@@ -0,0 +1,27 @@
+import { useState, useLayoutEffect } from 'react'
+
+import { isEqual } from 'lodash'
+
+/**
+ * Given a value and a function to determine equality, return a
+ * referentially equal value if the equality function returns true.
+ * This hook is helpful in avoiding re-renders and effects running
+ * based on an object or value that always changes references but
+ * infrequently changes it's value. By default, uses isEqual from
+ * lodash. This is typically only useful with objects and arrays.
+ *
+ * @param {T}                 value      - any value to check equality of
+ * @param {(T, T) => boolean} equalityFn - A function to determine equality
+ * @returns {T}
+ */
+export function useEqualityCheck (value, equalityFn = isEqual) {
+  const [computedValue, setComputedValue] = useState(() => value)
```

Comment:
Why would the default value be a function :thinking: Was this what was intended?
```suggestion
  const [computedValue, setComputedValue] = useState(value)
```

Reply:
Hah, I think I got lazy initialization happy with some of Dan's Swaps prs because the default values were being set by iterating through an array.

e.g
```js
const [state, setState] = useState(arrayFromProps.find(i => i.key === keyFromProps))
```

which would do that array iteration every time. So using the lazy evaluation by passing a function means the function only gets called on first render preventing the array loop on every render. It's definitely not saving anything here except allocating an additional function every render needlessly. Thanks for catching that.